mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -267,34 +267,37 @@ public ModSeq getHighestModSeq(Mailbox mailbox) throws MailboxException {
     public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        save(mailbox, addUidAndModseq(message, mailboxId))
-            .thenEmpty(indexTableHandler.updateIndexOnAdd(message, mailboxId))
-            .block();
+        block(addUidAndModseq(message, mailboxId)
+            .flatMap(Throwing.function(messageWithUidAndModSeq -> save(mailbox, messageWithUidAndModSeq)))
+            .thenEmpty(Mono.defer(() -> indexTableHandler.updateIndexOnAdd(message, mailboxId))));
         return message.metaData();
     }
 
-    private MailboxMessage addUidAndModseq(MailboxMessage message, CassandraId mailboxId) throws MailboxException {
+    private Mono<MailboxMessage> addUidAndModseq(MailboxMessage message, CassandraId mailboxId) {
         Mono<MessageUid> messageUidMono = uidProvider
             .nextUid(mailboxId)
             .switchIfEmpty(Mono.error(() -> new MailboxException("Can not find a UID to save " + message.getMessageId() + " in " + mailboxId)));
 
         Mono<ModSeq> nextModSeqMono = modSeqProvider.nextModSeq(mailboxId)
             .switchIfEmpty(Mono.error(() -> new MailboxException("Can not find a MODSEQ to save " + message.getMessageId() + " in " + mailboxId)));
 
-        try {
-            Mono.zip(messageUidMono, nextModSeqMono)
+        return Mono.zip(messageUidMono, nextModSeqMono)
                 .doOnNext(tuple -> {
                     message.setUid(tuple.getT1());
                     message.setModSeq(tuple.getT2());
                 })
-                .block();
+                .then(Mono.just(message));
+    }
+
+    private <T> T block(Mono<T> mono) throws MailboxException {
```

Comment:
Is there really an advantage using this wrap method over a call to block() at the end of the chain?

Reply:
In order to preserve MailboxException handling, yes!